source/ConditionsStore.cpp

```diff
@@ -475,3 +496,26 @@ const ConditionsStore::ConditionEntry *ConditionsStore::GetEntry(const string &n
 	// And otherwise we don't have a match.
 	return nullptr;
 }
+
+
+
+// Helper function to check if we can safely add a provider with the given name.
+bool ConditionsStore::VerifyProviderLocation(const string &name) const
+{
+	auto it = storage.upper_bound(name);
+	if(it == storage.begin())
+		return true;
+
+	--it;
+	// If the entry is already there, then it apparently was safe to add the entry.
+	if(name == it->first)
+		return true;
+
+	const ConditionEntry &ce = it->second;
+	if(ce.provider && ce.provider->isPrefixProvider && 0 == name.compare(0, ce.provider->name.length(), ce.provider->name))
```

Comment:
The comparison is with `ce.provider->name`, not with `it.first` or `ce.fullKey`. `ce.provider->name` should be set to `foo:`, even when we get it through the specific `foo: baa` entry, because `foo: baa` will be set to the `foo` provider when the `foo` provider was added. So the warning should still show.

I verified this with [an unit test](https://github.com/petervdmeer/endless-sky/blob/afbf0ccb3fed637c7f8804bacd1327bad179cb53/tests/unit/src/test_conditionsStore.cpp#L746) and execution of that test does show the error-message `Error: not adding provider for "ships: Large: ", because it is within range of prefixed derived provider "ships: "`.